Set expectedContentLength in ServiceWorkerDownloadTask didReceiveData callback

youennf opened the pull request:
#### 423d5f5dd0dbc7c48f3f1c931db7e5734fc692d5
<pre>
Set expectedContentLength in ServiceWorkerDownloadTask didReceiveData callback
<a href="https://bugs.webkit.org/show_bug.cgi?id=262297">https://bugs.webkit.org/show_bug.cgi?id=262297</a>
rdar://109266074

Reviewed by Alex Christensen.

Store the Response expectedContentLength in ServiceWorkerDownloadTask.
Use it to set the totalExpectedSize parameter of the didReceiveData callback.
We use the max of the expectedContentLength and the size of data already written as service worker downloads may use synthetic response with zero expectedContentLength.
For synthetic responses, if the expectedContentLength is zero, we check the Content-Length header and parse it to compute a new expectedContentLength value.

* LayoutTests/http/tests/workers/service/resources/service-worker-download-worker.js:
(event.event.request.url.includes):
* LayoutTests/http/tests/workers/service/service-worker-download-stream.https-expected.txt:
* LayoutTests/http/tests/workers/service/service-worker-download-stream.https.html:
* LayoutTests/http/wpt/service-workers/fetch-service-worker-preload-download-through-direct-preload.https-expected.txt:
* LayoutTests/http/wpt/service-workers/fetch-service-worker-preload-download-through-direct-preload.https.html:
* LayoutTests/platform/glib/http/tests/workers/service/service-worker-download-stream.https-expected.txt:
* LayoutTests/platform/glib/http/wpt/service-workers/fetch-service-worker-preload-download-through-direct-preload.https-expected.txt:
* LayoutTests/platform/ios/http/wpt/service-workers/fetch-service-worker-preload-download-through-direct-preload.https-expected.txt:
* Source/WebCore/Modules/fetch/FetchResponse.cpp:
(WebCore::FetchResponse::create):
* Source/WebKit/NetworkProcess/ServiceWorker/ServiceWorkerDownloadTask.cpp:
(WebKit::ServiceWorkerDownloadTask::ServiceWorkerDownloadTask):
(WebKit::ServiceWorkerDownloadTask::didReceiveData):
* Source/WebKit/NetworkProcess/ServiceWorker/ServiceWorkerDownloadTask.h:
(WebKit::ServiceWorkerDownloadTask::create):
* Source/WebKit/NetworkProcess/ServiceWorker/ServiceWorkerFetchTask.cpp:
(WebKit::ServiceWorkerFetchTask::convertToDownload):
* Tools/WebKitTestRunner/InjectedBundle/Bindings/TestRunner.idl:
* Tools/WebKitTestRunner/InjectedBundle/TestRunner.cpp:
(WTR::TestRunner::setShouldLogDownloadExpectedSize):
* Tools/WebKitTestRunner/InjectedBundle/TestRunner.h:
* Tools/WebKitTestRunner/TestController.cpp:
(WTR::TestController::resetStateToConsistentValues):
(WTR::TestController::downloadDidFinish):
(WTR::TestController::downloadDidWriteData):
* Tools/WebKitTestRunner/TestController.h:
(WTR::TestController::setShouldLogDownloadExpectedSize):
* Tools/WebKitTestRunner/TestInvocation.cpp:
(WTR::TestInvocation::didReceiveMessageFromInjectedBundle):

Canonical link: <a href="https://commits.webkit.org/268794@main">https://commits.webkit.org/268794@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/da734e22e19fc6b753c605ebfbd1f79c25d383bb

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/20684 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/21102 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/21758 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/22579 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/19307 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/20921 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/24357 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/21284 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/20638 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/20906 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/20733 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/17984 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/23435 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/17898 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/18790 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/25061 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/18977 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/18968 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/23012 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/19557 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/16618 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/18777 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/18626 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4968 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/23114 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/19375 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebKit/NetworkProcess/ServiceWorker/ServiceWorkerDownloadTask.h:
```diff
@@ -99,7 +99,8 @@ class ServiceWorkerDownloadTask : public NetworkDataTask, private FunctionDispat
     FileSystem::PlatformFileHandle m_downloadFile { FileSystem::invalidPlatformFileHandle };
-    int64_t m_downloadBytesWritten { 0 };
+    uint64_t m_downloadBytesWritten { 0 };
+    uint64_t m_expectedContentLength { 0 };
```
Shouldn't this be std::optional?  It's possible to not be there.

youennf replied:
0 for m_downloadBytesWritten means that didReceiveData was never called.
This is true for m_expectedContentLength.

achristensen07 commented on Tools/WebKitTestRunner/TestController.h:
```diff
@@ -533,7 +534,7 @@ class TestController {
     static void downloadDidReceiveAuthenticationChallenge(WKDownloadRef, WKAuthenticationChallengeRef, const void *clientInfo);
 
-    void downloadDidWriteData(long long totalBytesWritten);
+    void downloadDidWriteData(long long totalBytesWritten, long long totalBytesExpectedToWrite);
```
These should probably all be uint64_t instead of long long, but expected bytes should be std::optional<uint64_t>

achristensen07 commented on Source/WebKit/NetworkProcess/ServiceWorker/ServiceWorkerDownloadTask.cpp:
```diff
@@ -57,6 +58,11 @@ ServiceWorkerDownloadTask::ServiceWorkerDownloadTask(NetworkSession& session, Ne
     , m_downloadID(downloadID)
     , m_networkProcess(serviceWorkerConnection.networkProcess())
 {
+    m_expectedContentLength = response.expectedContentLength();
```
expectedContentLength() should probably be changed to return a std::optional<uint64_t>.  Its -1 value should be nullopt, but it was written before std::optional existed.  It shouldn't be stored as a std::optional, though, because that would increase the size of a common structure, ResourceResponseBase.

youennf replied:
Oh, I see why it is a int64_t now.
I can probably do this as a follow-up.

achristensen07 commented on Source/WebKit/NetworkProcess/ServiceWorker/ServiceWorkerDownloadTask.cpp:
```diff
@@ -57,6 +58,11 @@ ServiceWorkerDownloadTask::ServiceWorkerDownloadTask(NetworkSession& session, Ne
 {
+    m_expectedContentLength = response.expectedContentLength();
+    if (!m_expectedContentLength) {
+        if (auto contentLength = parseInteger<uint64_t>(response.httpHeaderField(HTTPHeaderName::ContentLength)))
```
This seems redundant.  Isn't this where the expected content length comes from?

youennf replied:
This covers the case of synthetic responses.
We could call  setExpectedContentLength at creation of the synthetic response instead.

youennf committed:
Set expectedContentLength in ServiceWorkerDownloadTask didReceiveData callback
https://bugs.webkit.org/show_bug.cgi?id=262297
rdar://109266074

Reviewed by Alex Christensen.

Store the Response expectedContentLength in ServiceWorkerDownloadTask.
Use it to set the totalExpectedSize parameter of the didReceiveData callback.
We use the max of the expectedContentLength and the size of data already written as service worker downloads may use synthetic response with zero expectedContentLength.
For synthetic responses, if the expectedContentLength is zero, we check the Content-Length header and parse it to compute a new expectedContentLength value.

* LayoutTests/http/tests/workers/service/resources/service-worker-download-worker.js:
(event.event.request.url.includes):
* LayoutTests/http/tests/workers/service/service-worker-download-stream.https-expected.txt:
* LayoutTests/http/tests/workers/service/service-worker-download-stream.https.html:
* LayoutTests/http/wpt/service-workers/fetch-service-worker-preload-download-through-direct-preload.https-expected.txt:
* LayoutTests/http/wpt/service-workers/fetch-service-worker-preload-download-through-direct-preload.https.html:
* LayoutTests/platform/glib/http/tests/workers/service/service-worker-download-stream.https-expected.txt:
* LayoutTests/platform/glib/http/wpt/service-workers/fetch-service-worker-preload-download-through-direct-preload.https-expected.txt:
* LayoutTests/platform/ios/http/wpt/service-workers/fetch-service-worker-preload-download-through-direct-preload.https-expected.txt:
* Source/WebCore/Modules/fetch/FetchResponse.cpp:
(WebCore::FetchResponse::create):
* Source/WebKit/NetworkProcess/ServiceWorker/ServiceWorkerDownloadTask.cpp:
(WebKit::ServiceWorkerDownloadTask::ServiceWorkerDownloadTask):
(WebKit::ServiceWorkerDownloadTask::didReceiveData):
* Source/WebKit/NetworkProcess/ServiceWorker/ServiceWorkerDownloadTask.h:
(WebKit::ServiceWorkerDownloadTask::create):
* Source/WebKit/NetworkProcess/ServiceWorker/ServiceWorkerFetchTask.cpp:
(WebKit::ServiceWorkerFetchTask::convertToDownload):
* Tools/WebKitTestRunner/InjectedBundle/Bindings/TestRunner.idl:
* Tools/WebKitTestRunner/InjectedBundle/TestRunner.cpp:
(WTR::TestRunner::setShouldLogDownloadExpectedSize):
* Tools/WebKitTestRunner/InjectedBundle/TestRunner.h:
* Tools/WebKitTestRunner/TestController.cpp:
(WTR::TestController::resetStateToConsistentValues):
(WTR::TestController::downloadDidFinish):
(WTR::TestController::downloadDidWriteData):
* Tools/WebKitTestRunner/TestController.h:
(WTR::TestController::setShouldLogDownloadExpectedSize):
* Tools/WebKitTestRunner/TestInvocation.cpp:
(WTR::TestInvocation::didReceiveMessageFromInjectedBundle):

Canonical link: https://commits.webkit.org/268794@main
